module-system/sov-modules-api/src/containers/kernel_value.rs

```diff
@@ -0,0 +1,396 @@
+use std::marker::PhantomData;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use sov_modules_core::{
+    Context, KernelWorkingSet, Prefix, StateCodec, StateKeyCodec, StateReaderAndWriter,
+    StateValueCodec, VersionedWorkingSet,
+};
+use sov_state::codec::BorshCodec;
+
+/// A `versioned` value stored in kernel state. The semantics of this type are different
+/// depending on the priveleges of the accessor. For a standard ("user space") interaction
+/// via a `VersionedWorkingSet`, only one version of this value is accessible. Inside the kernel,
+/// (where access is mediated by a `KernelWorkingSet`), all versions of this value are accessible.
+///
+/// Under the hood, a versioned value is implemented as a map from a slot number to a value. From the kernel, any
+/// value can be accessed using the `StateMapAcessor` trait with the slot number as the key. For convenience,
+/// the value can also be accessed using the `StateValueAccessor` trait, which will interact with the value for the current
+/// slot number.
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Clone,
+    BorshDeserialize,
+    BorshSerialize,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+pub struct KernelStateValue<V, Codec = BorshCodec> {
+    _phantom: PhantomData<V>,
+    codec: Codec,
+    prefix: Prefix,
+}
+
+impl<V> KernelStateValue<V> {
+    /// Crates a new [`KernelStateValue`] with the given prefix and the default
+    /// [`StateValueCodec`] (i.e. [`BorshCodec`]).
+    pub fn new(prefix: Prefix) -> Self {
+        Self::with_codec(prefix, BorshCodec)
+    }
+}
+
+impl<V, Codec> KernelStateValue<V, Codec> {
+    /// Creates a new [`KernelStateValue`] with the given prefix and codec.
+    pub fn with_codec(prefix: Prefix, codec: Codec) -> Self {
+        Self {
+            _phantom: PhantomData,
+            codec,
+            prefix,
+        }
+    }
+
+    /// Returns the prefix used when this [`KernelStateValue`] was created.
+    pub fn prefix(&self) -> &Prefix {
+        &self.prefix
+    }
+}
+
+mod as_user_value {
+    use super::*;
+    use crate::StateValueAccessor;
+
+    impl<'a, V, Codec, C: Context> StateValueAccessor<V, Codec, VersionedWorkingSet<'a, C>>
+        for KernelStateValue<V, Codec>
+    where
+        Codec: StateCodec,
+        Codec::ValueCodec: StateValueCodec<V>,
+        Codec::KeyCodec: StateKeyCodec<u64>,
+    {
+        fn prefix(&self) -> &Prefix {
+            &self.prefix
+        }
+
+        fn codec(&self) -> &Codec {
+            &self.codec
+        }
+
+        fn set(&self, value: &V, working_set: &mut VersionedWorkingSet<'a, C>) {
+            working_set.set_value(
+                self.prefix(),
+                &working_set.slot_num(),
+                value,
+                StateValueAccessor::<V, Codec, VersionedWorkingSet<'a, C>>::codec(self),
```

Comment:
Let's use `this::codec` instead of `StateValueAccessor::<V, Codec, VersionedWorkingSet<'a, C>>::codec(self)`

Reply:
Since this type implements two `Accessor` traits which each have a `codec` method, we have to disambiguate for rustc. Both impls are the same, but rustc doesn't know that :)